colcon_sanitizer_reports/sanitizer_log_parser.py

```diff
@@ -0,0 +1,195 @@
+from collections import defaultdict
+import csv
+from io import StringIO
+import re
+from typing import Dict, List, Optional, Tuple
+
+
+class _KeysFactory:
+    """Finds key stack trace lines in sanitizer error subsections."""
+
+    @staticmethod
+    def _make_masked_line(line: str) -> str:
+        """Masks numbers in a sanitizer error line so that keys from similar errors will match."""
+        masked_line = line
+        masked_line = re.sub(r'0x[\dabcdef]+', '0xX', masked_line)
+        masked_line = re.sub(r' \d+ ', ' X ', masked_line)
+        masked_line = re.sub(r'#\d+', '#X', masked_line)
+
+        return masked_line
+
+    @staticmethod
+    def _make_keys_base(*, lines: Tuple[str], stack_begin_regexes: Tuple[str, ...]) -> Tuple[str]:
+        keys: List[str] = []
+
+        lines = iter(lines)
+        for stack_begin_regex in stack_begin_regexes:
+            for line in lines:
+                match = re.match(
+                    stack_begin_regex,
+                    _KeysFactory._make_masked_line(line)
+                )
+                if match is not None:
+                    break
+
+            for line in lines:
+                match = re.match(
+                    r'^\s*#X (0xX in|)\s*(?P<key>.*/ros2.*)\s*$',
+                    _KeysFactory._make_masked_line(line)
+                )
+                if match is not None:
+                    keys.append(match.groupdict()['key'])
+                    break
+            else:
+                # We didn't find one of the keys we were required to find.
+                keys = []
+                break
+
+        return tuple(keys)
+
+    @staticmethod
+    def _make_default_keys(*, lines: Tuple[str]) -> Tuple[str]:
```

Comment:
Why using * here?

Reply:
This is pretty common in the colcon internal APIs to consume all positional arguments and force the usage of keyword arguments.